text/0201-scope-attributes.md

```diff
@@ -0,0 +1,138 @@
+# Introduce Scope Attributes
+
+This OTEP adds attributes to the Scope.
+
+## Motivation
+
+There are a few reasons why adding Scope attributes is a good idea:
+
+- There are 2 known use cases where Scope attributes can solve specific problems:
+  - Add support for [Meter "short_name"](https://github.com/open-telemetry/opentelemetry-specification/pull/2422),
+    represented as an attribute of Meter's Scope.
+  - Add support for differentiating the type of data emitted from the scopes that belong
+    to different data domains, e.g. profiling data emitted as log records or client-side
+    data emitted as log records needs to be differentiated so that it can be easily
+    routed and processed differently in the backends. We don't have a good way to handle
+    this today. The type of the data can be recorded as an attribute Logger's Scope.
+- It makes Scope consistent with the other primary data types: Resource, Span, Metric,
+  LogRecord.
+
+See additional [discussion here](https://github.com/open-telemetry/opentelemetry-specification/issues/2450).
+
+## Summary
+
+The following is the summary of proposed changes:
+
+- We will extend OpenTelemetry API to allow specifying Scope attributes when obtaining a
+  Tracer, Meter or LogEmitter. Scope attributes will be optional.
+- We will add `attributes` field to the [InstrumentationScope](https://github.com/open-telemetry/opentelemetry-proto/blob/88faab1197a2a105c7da659951e94bc951d37ab9/opentelemetry/proto/common/v1/common.proto#L83)
+  message of OTLP.
+- We will specify that Telemetry emitted via a Scope-ed Tracer, Meter or LogEmitter will
+  be associated with the Scope's attributes.
+- We will specify that OTLP Exporter will record the attributes in the
+  InstrumentationScope message.
+- We will create a section for Scope attributes' semantic conventions in
+  the specification.
+
+## Internal details
+
+### API Changes
+
+`Get a Tracer` API will be extended to add the following parameter:
+
+```
+- `attributes` (optional): Specifies the instrumentation scope attributes to associate
+  with emitted telemetry.
+```
+
+`Get a Metter` API will be extended to add the following parameter:
+
+```
+- `attributes` (optional): Specifies the instrumentation scope attributes to associate
+  with emitted telemetry.
+```
+
+`Get LogEmitter` SDK call will be altered to the following:
+
+```
+Accepts the instrumentation scope name and optional version and attributes and
+returns a LogEmitter associated with the instrumentation scope.
+```
+
+Since the attributes are optional this is a backwards compatible change.
```

Comment:
There might be value in making the attributes mutable, so that a common `state` can be recorded. The scope identifying fields are separate from attributes anyway.

Reply:
I am not sure I understand the mutability reference and common state can be recorded. Please elaborate.

Reply:
Something like a client side session-id, if we decide to use at the level of instrumentation scope, will be changing over time. Hence it will need this ability for the attributes to change.

Reply:
>Hence it will need this ability for the attributes to change.

That's one possible approach. Another approach is to get a new Scope every time you have a new session-id.

I don't know what's a better way. In this OTEP I thought it is going to be immutable, but I am open to re-considering.